internal/bgp/bgpfrr/frr_test.go

```diff
@@ -0,0 +1,333 @@
+
+	testCheckConfigFile(t)
+	session.Close()
+}
+
+func TestSingleSessionClose(t *testing.T) {
+	testSetup(t)
+
+	l := log.NewNopLogger()
+	session, err := New(l, "10.2.2.254", net.IPv4(10, 1, 1, 254), 100, net.IPv4(10, 1, 1, 254), 200, time.Second, "password", "hostname")
+	if err != nil {
+		t.Fatalf("Could not create session: %s", err)
+	}
+
+	session.Close()
+	testCheckConfigFile(t)
+}
+func TestTwoSessions(t *testing.T) {
+	testSetup(t)
+
+	l := log.NewNopLogger()
+	session1, err := New(l, "10.2.2.254", net.IPv4(10, 1, 1, 254), 100, net.IPv4(10, 1, 1, 254), 200, time.Second, "password", "hostname")
+	if err != nil {
+		t.Fatalf("Could not create session: %s", err)
+	}
+	session2, err := New(l, "10.4.4.254", net.IPv4(10, 3, 3, 254), 300, net.IPv4(10, 3, 3, 254), 400, time.Second, "password", "hostname")
+	if err != nil {
+		t.Fatalf("Could not create session: %s", err)
+	}
+
+	testCheckConfigFile(t)
+	session1.Close()
+	session2.Close()
+}
+
+func TestTwoSessionsDuplicate(t *testing.T) {
+	testSetup(t)
+
+	l := log.NewNopLogger()
+	session1, err := New(l, "10.2.2.254", net.IPv4(10, 1, 1, 254), 100, net.IPv4(10, 1, 1, 254), 200, time.Second, "password", "hostname")
+	if err != nil {
+		t.Fatalf("Could not create session: %s", err)
+	}
+	session2, err := New(l, "10.2.2.254", net.IPv4(10, 1, 1, 254), 100, net.IPv4(10, 1, 1, 254), 200, time.Second, "password", "hostname")
+	if err != nil {
+		t.Fatalf("Could not create session: %s", err)
+	}
+
+	testCheckConfigFile(t)
+	session1.Close()
+	session2.Close()
+}
+
+func TestTwoSessionsDuplicateRouter(t *testing.T) {
+	testSetup(t)
+
+	l := log.NewNopLogger()
+	session1, err := New(l, "10.2.2.254", net.IPv4(10, 1, 1, 254), 100, net.IPv4(10, 1, 1, 254), 200, time.Second, "password", "hostname")
+	if err != nil {
+		t.Fatalf("Could not create session: %s", err)
+	}
+	session2, err := New(l, "10.4.4.254", net.IPv4(10, 1, 1, 254), 100, net.IPv4(10, 1, 1, 254), 400, time.Second, "password", "hostname")
+	if err != nil {
+		t.Fatalf("Could not create session: %s", err)
+	}
+
+	testCheckConfigFile(t)
+	session1.Close()
+	session2.Close()
+}
+
+func TestSingleAdvertisement(t *testing.T) {
+	testSetup(t)
+
+	l := log.NewNopLogger()
+	session, err := New(l, "10.2.2.254", net.IPv4(10, 1, 1, 254), 100, net.IPv4(10, 1, 1, 254), 200, time.Second, "password", "hostname")
+	if err != nil {
+		t.Fatalf("Could not create session: %s", err)
+	}
+
+	prefix := &net.IPNet{
+		IP:   net.IPv4(172, 16, 1, 10),
+		Mask: net.IPv4Mask(255, 255, 255, 0),
+	}
+
+	adv := &bgp.Advertisement{
+		Prefix:  prefix,
+		NextHop: net.IPv4(10, 1, 1, 1),
+	}
+
+	err = session.Set(adv)
+	if err != nil {
+		t.Fatalf("Could not advertise prefix: %s", err)
+	}
+
+	testCheckConfigFile(t)
+	session.Close()
+}
+
+func TestSingleAdvertisementInvalidPrefix(t *testing.T) {
+	testSetup(t)
+
+	l := log.NewNopLogger()
+	session, err := New(l, "10.2.2.254", net.IPv4(10, 1, 1, 254), 100, net.IPv4(10, 1, 1, 254), 200, time.Second, "password", "hostname")
```

Comment:
You can use `net.ParseIP`

Reply:
You mean instead of `net.IPv4`? Ok, I can do that. I think I just copied this pattern from somewhere else in MetalLB.

Reply:
Seems more readable than thinking about net.IPV4 being a constructor